### ISSUE TITLE
subs is not working in logic expression

### ISSUE THREAD

**Issue by xurq5953** [sympy/sympy#21971]
subs is not working in logic expression, for a long expression it only replacement the first match subexpression

**Comment by sidhu1012** [comment]
It's better to share code snippet in issue.

**Comment by xurq5953** [comment]
```
fun1 = g ^ h & g ^ h & f ^ g & e ^ h & g & e ^ d ^ h & g & d ^ f & d ^ h & f & d ^ g & f & d ^ h & g & f & d ^ e & d ^ g & e & d ^ h & f & e & d ^ g & f & e & d ^ g & c ^ h & g & c \
     ^ f & c ^ h & f & c ^ g & f & c ^ e & c ^ h & e & c ^ g & e & c ^ f & e & c ^ h & f & e & c ^ h & g & f & e & c ^ d & c ^ h & d & c ^ h & g & d & c ^ f & d & c ^ h & f & d & c ^ g & \
     f & d & c ^ h & g & f & d & c ^ h & e & d & c ^ g & e & d & c ^ f & e & d & c ^ h & f & e & d & c ^ h & g & f & e & d & c ^ b ^ g & b ^ h & f & b ^ h & g & e & b ^ f & e & b ^ h & f & e & b \
     ^ h & g & f & e & b ^ h & d & b ^ g & d & b ^ h & f & d & b ^ g & f & d & b ^ h & g & f & d & b ^ e & d & b ^ h & g & e & d & b ^ h & f & e & d & b ^ g & f & e & d & b ^ c & b ^ g & c & b ^ \
     g & f & c & b ^ g & e & c & b ^ f & e & c & b ^ h & g & f & e & c & b ^ g & d & c & b ^ h & g & d & c & b ^ g & f & d & c & b ^ h & g & f & d & c & b ^ e & d & c & b ^ g & e & d & c & b ^ h & g & e & \
     d & c & b ^ f & e & d & c & b ^ h & f & e & d & c & b ^ h & g & f & e & d & c & b ^ h & a ^ g & a ^ h & f & a ^ h & g & f & a ^ e & a ^ h & e & a ^ g & f & e & a ^ h & g & f & e & a ^ d & a \
     ^ h & d & a ^ g & d & a ^ h & g & d & a ^ g & f & d & a ^ h & g & f & d & a ^ e & d & a ^ h & e & d & a ^ f & e & d & a ^ h & f & e & d & a ^ g & f & e & d & a ^ h & g & f & e & d & a ^ c & a \
     ^ h & g & c & a ^ e & c & a ^ h & g & e & c & a ^ f & e & c & a ^ h & f & e & c & a ^ h & g & f & e & c & a ^ d & c & a ^ h & d & c & a ^ e & d & c & a ^ g & e & d & c & a ^ f & e & d & c & a ^ \
     h & f & e & d & c & a ^ h & b & a ^ g & b & a ^ h & f & b & a ^ g & f & b & a ^ h & g & f & b & a ^ e & b & a ^ h & e & b & a ^ f & e & b & a ^ d & b & a ^ h & d & b & a ^ g & d & b & a ^ h & \
     g & d & b & a ^ h & f & d & b & a ^ g & f & d & b & a ^ h & g & f & d & b & a ^ h & e & d & b & a ^ f & e & d & b & a ^ h & f & e & d & b & a ^ h & g & f & e & d & b & a ^ h & f & c & b & a ^ h & e & c & \
     b & a ^ g & e & c & b & a ^ g & f & e & c & b & a ^ h & d & c & b & a ^ g & d & c & b & a ^ h & g & d & c & b & a ^ f & d & c & b & a ^ g & f & d & c & b & a ^ g & e & d & c & b & a ^ h & f & e & d & c & \
     b & a ^ g & f & e & d & c & b & a

fun1.subs(b & c, P)
```

It only subs the first match one , for replace() is the same.

**Comment by oscargus** [comment]
I can confirm it. Here is a simpler example:

```
from sympy import *
a, b, c, d = symbols('a b c d')
f = a & b & c | a & c
f.subs(a & c, d)
```
returns
```
d | (a & b & c)
```

To compare with 
```
f2 = a * b * c + a * c
f2.subs(a * c, d)
```
which returns
```
b*d + d
```

**Comment by oscargus** [comment]
Probably related to #14836 (in the sense that if one is solved, probably so is the other although not the same problem).

**Comment by oscargus** [comment]
I added a fix in #21976 that should work for `And`, `Or`, and `Xor`.
